instance/migrations/0042_add_instance_status.py

```diff
@@ -6,40 +6,37 @@
 from instance.models.instance import Status as InstanceStatus
 
 
+def get_current_server(instance):
+    return instance.server_set.order_by("id").last()
+
+def get_instance_state(server_status, server_progress):
+    if server_status in ("new", "active"):
+        return InstanceStatus.WaitingForServer.state_id
+    if server_status == "started":
+        if server_progress in ("running", "success"):
+            return InstanceStatus.WaitingForServer.state_id
+        elif server_progress == "failed":
+            return InstanceStatus.Error.state_id
+    if server_status in ("booted", "ready"):
+        return InstanceStatus.Running.state_id
+    if server_status == "provisioning":
+        if server_progress in ("running", "success"):
+            return InstanceStatus.ConfiguringServer.state_id
+        elif server_progress == "failed":
+            return InstanceStatus.ConfigurationFailed.state_id
+    if server_status == "rebooting":
+        return InstanceStatus.ConfiguringServer.state_id
+    if server_status == "terminated":
+        return InstanceStatus.Terminated.state_id
+
 def set_instance_states(apps, schema_editor):
     SingleVMOpenEdXInstance = apps.get_model("instance", "SingleVMOpenEdXInstance")
-    SingleVMOpenEdXInstance.objects.filter(
-        server_set___status__in=("new", "active")).update(
-            _status=InstanceStatus.WaitingForServer.state_id
-        )
-    SingleVMOpenEdXInstance.objects.filter(
-        server_set___status="started", server_set___progress__in=("running", "success")).update(
-            _status=InstanceStatus.WaitingForServer.state_id
-        )
-    SingleVMOpenEdXInstance.objects.filter(
-        server_set___status="started", server_set___progress="failed").update(
-            _status=InstanceStatus.Error.state_id
-        )
-    SingleVMOpenEdXInstance.objects.filter(
-        server_set___status__in=("booted", "ready")).update(
-            _status=InstanceStatus.Running.state_id
-        )
-    SingleVMOpenEdXInstance.objects.filter(
-        server_set___status="provisioning", server_set___progress__in=("running", "success")).update(
-            _status=InstanceStatus.ConfiguringServer.state_id
-        )
-    SingleVMOpenEdXInstance.objects.filter(
-        server_set___status="provisioning", server_set___progress="failed").update(
-            _status=InstanceStatus.ConfigurationFailed.state_id
-        )
-    SingleVMOpenEdXInstance.objects.filter(
-        server_set___status="rebooting").update(
-            _status=InstanceStatus.ConfiguringServer.state_id
-        )
-    SingleVMOpenEdXInstance.objects.filter(
-        server_set___status="terminated").update(
-            _status=InstanceStatus.Terminated.state_id
-        )
+    for instance in SingleVMOpenEdXInstance.objects.iterator():
+        current_server = get_current_server(instance)
+        server_status = current_server._status
```

Comment:
Getting an error here when there is no server attached to an instance (which happens when there are more instances created than workers are available to spawn servers):

```
  Applying instance.0042_add_instance_status...Traceback (most recent call last):
  File "./manage.py", line 37, in <module>
    execute_from_command_line(sys.argv)
  File "/home/vagrant/.virtualenvs/opencraft/lib/python3.4/site-packages/django/core/management/__init__.py", line 351, in execute_from_command_line
    utility.execute()
  File "/home/vagrant/.virtualenvs/opencraft/lib/python3.4/site-packages/django/core/management/__init__.py", line 343, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/vagrant/.virtualenvs/opencraft/lib/python3.4/site-packages/django/core/management/base.py", line 394, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/vagrant/.virtualenvs/opencraft/lib/python3.4/site-packages/django/core/management/base.py", line 445, in execute
    output = self.handle(*args, **options)
  File "/home/vagrant/.virtualenvs/opencraft/lib/python3.4/site-packages/django/core/management/commands/migrate.py", line 222, in handle
    executor.migrate(targets, plan, fake=fake, fake_initial=fake_initial)
  File "/home/vagrant/.virtualenvs/opencraft/lib/python3.4/site-packages/django/db/migrations/executor.py", line 110, in migrate
    self.apply_migration(states[migration], migration, fake=fake, fake_initial=fake_initial)
  File "/home/vagrant/.virtualenvs/opencraft/lib/python3.4/site-packages/django/db/migrations/executor.py", line 148, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/home/vagrant/.virtualenvs/opencraft/lib/python3.4/site-packages/django/db/migrations/migration.py", line 115, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/home/vagrant/.virtualenvs/opencraft/lib/python3.4/site-packages/django/db/migrations/operations/special.py", line 183, in database_forwards
    self.code(from_state.apps, schema_editor)
  File "/vagrant/instance/migrations/0042_add_instance_status.py", line 36, in set_instance_states
    server_status = current_server._status
AttributeError: 'NoneType' object has no attribute '_status'
```

Just needs an `if current_server:` and to handle setting the status to `New` in that case.